base/shell/explorer/notifyiconscust.cpp

```diff
@@ -20,26 +20,194 @@
+{
+    LVITEMW iItem;
+
+    iItem.mask = LVIF_TEXT | LVIF_IMAGE;
+    iItem.pszText = data->text;
+    iItem.cchTextMax = lstrlenW(data->text);
+    iItem.iSubItem = 0;
+    iItem.iItem = id;
+    iItem.iImage = id;
+
+    if (ListView_InsertItem(hListView, &iItem) == -1)
+        return FALSE;
+    else
+        return TRUE;
+}
+
+// Return Toolbar Window hander on successful, Null otherwise
+HWND FindTrayToolbarWindowHandle()
+{
+    HWND hwnd;
+    if ((hwnd = FindWindowW(L"Shell_TrayWnd", NULL)))
+        if ((hwnd = FindWindowExW(hwnd, NULL, L"TrayNotifyWnd", NULL)))
+            if ((hwnd = FindWindowExW(hwnd, NULL, L"SysPager", NULL)))
+                if ((hwnd = FindWindowExW(hwnd, NULL, L"ToolbarWindow32", NULL)))
+                    return hwnd;
+
+    return NULL;
+}
+
+BOOL ExtractData(HWND hListView, HIMAGELIST hImageList)
+{
+    const int BUFFER_SIZE = 0x1000;
+    VOID* pLocalBuffer = (VOID*)new CHAR[BUFFER_SIZE];
+
+    HWND hToolbar = FindTrayToolbarWindowHandle();
+    int count = SendMessageW(hToolbar, TB_BUTTONCOUNT, 0, 0);
+
+    DWORD processId = 0;
+    GetWindowThreadProcessId(hToolbar, &processId);
+    HANDLE hProcess = OpenProcess(PROCESS_ALL_ACCESS, FALSE, processId);
+    if (!hProcess)
+        return FALSE;
+
+    VOID *pRemoteBuffer = VirtualAllocEx(hProcess, NULL, BUFFER_SIZE, MEM_COMMIT, PAGE_READWRITE);
+    if (!pRemoteBuffer)
+        return FALSE;
+
+    ZeroMemory(pLocalBuffer, BUFFER_SIZE);
+
+    hImageList = ImageList_Create(BITMAP_WIDTH, BITMAP_HEIGHT, FALSE, count, 0);
+
+    for (int i = 0; i < count; i++)
+    {
+        DWORD pBytesRead;
+
+        // TBBUTTON
+        if (SendMessageW(hToolbar, TB_GETBUTTON, i, (LPARAM)pRemoteBuffer))
+        {
+            pBytesRead = 0;
+            if (ReadProcessMemory(hProcess, pRemoteBuffer, pLocalBuffer, sizeof(TBBUTTON), &pBytesRead))
+            {
+         //  dwData
+                pBytesRead = 0;
+                TBBUTTON *tbb = reinterpret_cast<TBBUTTON*>(pLocalBuffer);
+                TRAYDATA *pData = reinterpret_cast<TRAYDATA*>(((CHAR*)pLocalBuffer) + sizeof(TBBUTTON));
+                if (ReadProcessMemory(hProcess, (VOID*)tbb->dwData, pData, sizeof(TRAYDATA), &pBytesRead))
+                {
+                    HICON hIcon;
+                    if ((hIcon = CopyIcon(pData->hIcon)))
+                        ImageList_AddIcon(hImageList, hIcon);
+
+                    InsertItem(hListView, tbb, pData, i);
+                }
+            }
+        }
+    }
+    ListView_SetImageList(hListView, hImageList, LVSIL_SMALL);
+
+    VirtualFree(pRemoteBuffer, 0, MEM_RELEASE);
+    CloseHandle(hProcess);
+
+    delete (CHAR*)pLocalBuffer;
+    return TRUE;
+}
+
 INT_PTR CALLBACK CustomizeNotifyIconsProc(HWND hwnd, UINT Message, WPARAM wParam, LPARAM lParam)
 {
-    switch(Message)
+    static HWND hListView = NULL;
+    static HWND hCombo = NULL;
+    static HIMAGELIST hImageList;
```

Comment:
static? So what happens if the window procedure is called by two different threads???